rust/feature-flags/src/flags/flag_matching.rs

```diff
@@ -425,62 +426,192 @@ impl FeatureFlagMatcher {
+        );
+
+        inc(
+            FLAG_EVALUATION_ERROR_COUNTER,
+            &[("reason".to_string(), reason.to_string())],
+            1,
+        );
+    }
+
+    /// Evaluates flags using dependency graph to handle flag dependencies correctly.
+    /// Returns true if there were errors during evaluation.
+    async fn evaluate_flags_with_dependency_graph(
+        &mut self,
+        feature_flags: &FeatureFlagList,
+        person_property_overrides: &Option<HashMap<String, Value>>,
+        group_property_overrides: &Option<HashMap<String, HashMap<String, Value>>>,
+        hash_key_overrides: &Option<HashMap<String, String>>,
+        evaluated_flags_map: &mut HashMap<String, FlagDetails>,
+    ) -> bool {
+        let (flag_dependency_graph, errors) =
+            match DependencyGraph::from_nodes(&feature_flags.flags) {
+                Ok((graph, errors)) => (graph, errors),
+                Err(e) => {
+                    self.handle_dependency_graph_error("build feature flag dependency graph", &e);
+                    return true;
+                }
+            };
+
+        let mut errors_while_computing_flags = !errors.is_empty();
+        if errors_while_computing_flags {
+            self.handle_dependency_graph_errors(&errors);
+        }
+
+        let evaluation_stages = match flag_dependency_graph.evaluation_stages() {
+            Ok(stages) => stages,
+            Err(e) => {
+                self.handle_dependency_graph_error("get evaluation stages", &e);
+                debug_assert!(
+                    false,
+                    "evaluation_stages() failed after dependency graph construction: {e:?}"
+                );
+                return true;
+            }
+        };
+
+        // Evaluate flags in dependency order
+        for stage in evaluation_stages {
+            let stage_flags: Vec<FeatureFlag> = stage.iter().map(|&flag| flag.clone()).collect();
+            let (level_evaluated_flags_map, level_errors) = self
+                .evaluate_flags_in_level(
+                    &stage_flags,
+                    evaluated_flags_map,
+                    person_property_overrides,
+                    group_property_overrides,
+                    hash_key_overrides,
+                )
+                .await;
+            errors_while_computing_flags |= level_errors;
+            evaluated_flags_map.extend(level_evaluated_flags_map);
         }
+
+        errors_while_computing_flags
+    }
+
+    /// Handles errors during dependency graph operations.
+    fn handle_dependency_graph_error(&self, error_type: &str, error: &dyn std::fmt::Debug) {
+        error!(
+            "Failed to {} for team {}: {:?}",
+            error_type, self.team_id, error
+        );
+        inc(
+            FLAG_EVALUATION_ERROR_COUNTER,
+            &[(
+                "reason".to_string(),
+                format!("{}_error", error_type.replace(" ", "_")),
```

Comment:
nit: Unless someone is already aware of this formatting convention, it'll be hard to figure out where `build_feature_flag_dependency_graph_error` is coming from. They will search the code base for `build_feature_flag_dependency_graph_error`, and it will return no results, which is a bit painful, especially during an incident.

Reply:
oh wait this is a metric, ignore me